### ISSUE TITLE
[Homework5] counter_sleep() case fail, ABA problem 

### ISSUE THREAD

**Issue by quanxi1** [kaist-cp/cs431#861]
Hello TAs, 

I've passed all the other test cases except for the counter_sleep().

After debugging and conferring to several previous questions which were about similar problems, I'm convinced that there is a problem in protecting a pointer (preventing memory reclamation).

I assume that the returned hazard pointer is not protecting it perfectly. During the iteration, time to time, `cur_ptr` shows weird values and I guess that this is because `collect()` is freeing even protected pointers and thus use-after-free happens. 

I thoroughly checked my implementation for Shield methods (set, protect, try_protect, validate) and HazardBag methods, but I think my current code properly takes into account all the specifications described; in particular the part below :
```

(T1-1) Add b to the hazard list       | (T2-1) Unlink b and `retire(b)`
       (`Shield::try_protect()`)      |
(T1-2) Check if b is still reachable  | (T2-2) Check if b is in the hazard list
       if so, deref b                 |        if not, free b
(T1-3) Remove b from the hazard list  |
       (`Shield::drop()`)             |
```

I found out that some of the previous questions were talking about ABA problem, but I couldn't understand why ABA problem takes place here. Is the ABA problem here indicating that in `validate()`, even though `*src == pointer` has passed, it can be possibly the case that `*src` had been changed but eventually became `pointer`? 
Is it even possible in our test case? Because each thread is creating a new Box in each iteration, I thought that the CASed new_ptr values is always new so such hazard will never happen(at least in this test case). If it actually becomes probem, how can I resolve it? 

So in conclusion, I want to get help in 

1. What exactly does the 'ABA problem' mean in this context? 
2. Could you give me any advice about to which part I should pay attention to resolve current problem?

**Comment by quanxi1** [comment]
Sorry I put wrong tag and I can't modify it :(

**Comment by quanxi1** [comment]
These are the previous discussions I referred to : 
https://github.com/kaist-cp/cs431/issues/706#issue-1436836815 
https://github.com/kaist-cp/cs431/issues/560#issue-1071228116
https://github.com/kaist-cp/cs431/issues/581#issue-1077547800

**Comment by tomtomjhj** [comment]
ABA is just one consequence of the incorrect implementation. #560 was about a bug in the old version of the template code, which has been fixed. 

I've seen some implementation where "Add b to the hazard list" part is subtly wrong (#706, #790).
